omniscidb/ResultSetRegistry/ResultSetRegistry.cpp

```diff
@@ -0,0 +1,267 @@
+      check_init_lock.unlock();
+      mapd_unique_lock<mapd_shared_mutex> init_lock(res_registry_init_mutex);
+      if (!ps_mgr->hasDataProvider(hdk::ResultSetRegistry::SCHEMA_ID)) {
+        ps_mgr->registerDataProvider(hdk::ResultSetRegistry::SCHEMA_ID,
+                                     std::make_shared<hdk::ResultSetRegistry>(config));
+      }
+    }
+  }
+
+  auto provider = ps_mgr->getDataProvider(hdk::ResultSetRegistry::SCHEMA_ID);
+  auto res = std::dynamic_pointer_cast<hdk::ResultSetRegistry>(provider);
+  CHECK(res);
+  return res;
+}
+
+ResultSetTableTokenPtr ResultSetRegistry::put(ResultSetTable table) {
+  CHECK(!table.empty());
+
+  mapd_unique_lock<mapd_shared_mutex> schema_lock(schema_mutex_);
+  mapd_unique_lock<mapd_shared_mutex> data_lock(data_mutex_);
+
+  auto table_id = next_table_id_++;
+  // Add schema information for the ResultSet.
+  auto tinfo = addTableInfo(db_id_,
+                            table_id,
+                            ResultSetTableToken::tableName(table_id),
+                            false,
+                            Data_Namespace::MemoryLevel::CPU_LEVEL,
+                            table.size());
+  auto& first_rs = table.result(0);
+  for (size_t col_idx = 0; col_idx < first_rs->colCount(); ++col_idx) {
+    addColumnInfo(db_id_,
+                  table_id,
+                  static_cast<int>(col_idx + 1),
+                  first_rs->colName(col_idx),
+                  first_rs->colType(col_idx),
+                  false);
+  }
+
+  auto table_data = std::make_unique<TableData>();
+  size_t row_count = 0;
+  for (auto& rs : table.results()) {
+    DataFragment frag;
+    frag.offset = row_count;
+    frag.row_count = rs->rowCount();
+    frag.rs = rs;
+    if (useColumnarResults(*rs)) {
+      frag.mutex = std::make_unique<std::mutex>();
+    }
+
+    row_count += frag.row_count;
+    table_data->fragments.emplace_back(std::move(frag));
+  }
+  table_data->row_count = row_count;
+
+  tables_[table_id] = std::move(table_data);
+
+  return std::make_shared<ResultSetTableToken>(tinfo, row_count, shared_from_this());
+}
+
+ResultSetPtr ResultSetRegistry::get(const ResultSetTableToken& token, size_t idx) const {
+  mapd_shared_lock<mapd_shared_mutex> data_lock(data_mutex_);
+  CHECK(tables_.count(token.tableId()));
+  auto* table = tables_.at(token.tableId()).get();
+  mapd_shared_lock<mapd_shared_mutex> table_lock(table->mutex);
+  CHECK_LT(idx, table->fragments.size());
+  return table->fragments[idx].rs;
+}
+
+void ResultSetRegistry::drop(const ResultSetTableToken& token) {
+  mapd_unique_lock<mapd_shared_mutex> schema_lock(schema_mutex_);
+  mapd_unique_lock<mapd_shared_mutex> data_lock(data_mutex_);
+
+  CHECK(tables_.count(token.tableId()));
+  std::unique_ptr<TableData> table = std::move(tables_.at(token.tableId()));
+  mapd_unique_lock<mapd_shared_mutex> table_lock(table->mutex);
+  tables_.erase(token.tableId());
```

Comment:
Do I understand correctly that you hold data_mutex_ here to prevent map rebalancing? Would it be possible to change the container (to limit pointers invalidation) and narrow the lock scope? I'm not sure it's too important, just a note.

Reply:
I don't think it matters much which exact structure we use for `tables_`. We would still need a lock to read/write this structure. And the fact that this map holds `unique_ptr` with table data allows us to release the lock once table data is found. So I usually do `data_lock.unlock();` when the required table pointer is found. So the scope of this lock shouldn't be too big.